### PR TITLE
Replace virtual functions with static_cast

### PR DESCRIPTION
**Pull request by ramon-bernardo**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
- Change for static_cast on conversion is guaranteed.

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

File: src/monster.cpp [hunk 1]
```diff
@@ -90,9 +90,7 @@ void Monster::setName(const std::string& name)
 	SpectatorVec spectators;
 	g_game.map.getSpectators(spectators, position, true, true);
 	for (Creature* spectator : spectators) {
-		if (Player* tmpPlayer = spectator->getPlayer()) {
-			tmpPlayer->sendUpdateTileCreature(this);
-		}
```
Comment (ramon-bernardo) [inline src/monster.cpp:95]: This is not necessary because `getSpectators - onlyPlayers` is true

File: src/condition.cpp [hunk 2]
```diff
@@ -915,7 +915,7 @@ bool ConditionRegeneration::executeCondition(Creature* creature, int32_t interva
 					message.type = MESSAGE_HEALED_OTHERS;
 					message.text = player->getName() + " was healed for " + healString;
 					for (Creature* spectator : spectators) {
-						spectator->getPlayer()->sendTextMessage(message);
+						static_cast<Player*>(spectator)->sendTextMessage(message);
```
Comment (ranisalt) [inline src/condition.cpp:918]: Please add `assert(dynamic_cast<Player*>(spectator) != nullptr)` before so we catch bugs on debug builds

**Review by ranisalt — CHANGES_REQUESTED** [review]

File: src/monster.cpp [hunk 3]
```diff
@@ -479,7 +477,7 @@ bool Monster::isOpponent(const Creature* creature) const
 			return true;
 		}
 	} else {
-		if ((creature->getPlayer() && !creature->getPlayer()->hasFlag(PlayerFlag_IgnoredByMonsters)) ||
+		if ((creature->getPlayer() && !static_cast<const Player*>(creature)->hasFlag(PlayerFlag_IgnoredByMonsters)) ||
```
Comment (ranisalt) [inline src/monster.cpp:480]: I think this is better:
```suggestion
		if (Player* player = creature->getPlayer(); player && !player->hasFlag(PlayerFlag_IgnoredByMonsters)) ||
```

**Comment by ranisalt** [comment]
You'll need to rebase and reformat the files

**Comment by Zbizu** [comment]
what is the context for this change? performance gain? resolving potential crashes? code cleanup?

**Comment by ranisalt** [comment]
> what is the context for this pull request? performance gain? resolving potential crashes? code cleanup?

It saves one virtual function call - which cannot be inlined because the function address cannot be determined in compile time, while `static_cast` is free - when we are sure that the conversion is valid.

**Comment by DSpeichert** [comment]
Is this alive? Rebase needed.

File: src/game.cpp [hunk 4]
```diff
@@ -958,7 +958,7 @@ void Game::playerMoveItem(Player* player, const Position& fromPos,
 	}
 
 	const Position& playerPos = player->getPosition();
-	const Position& mapFromPos = fromCylinder->getTile()->getPosition();
+	const Position& mapFromPos = static_cast<Tile*>(fromCylinder)->getPosition();
```
Comment (ramon-bernardo) [inline src/game.cpp:961]: Review

File: src/player.cpp [hunk 5]
```diff
@@ -3103,7 +3105,7 @@ void Player::postAddNotification(Thing* thing, const Cylinder* oldParent, int32_
 		assert(i ? i->getContainer() != nullptr : true);
 
 		if (i) {
-			requireListUpdate = i->getContainer()->getHoldingPlayer() != this;
+			requireListUpdate = static_cast<const Container*>(i)->getHoldingPlayer() != this;
```
Comment (ramon-bernardo) [inline src/player.cpp:3108]: Review

File: src/player.cpp [hunk 6]
```diff
@@ -3159,7 +3161,7 @@ void Player::postRemoveNotification(Thing* thing, const Cylinder* newParent, int
...
```
Comment (ramon-bernardo) [inline src/player.cpp:3164]: Review

**Review by ramon-bernardo — COMMENTED** [review]
This needs attention

**Comment by ramon-bernardo** [comment]
The name of last commit is "Rename tmpPlayer to spectatorPlayer".
VSCode driving me crazy. 🤪

**Review by EPuncker — APPROVED** [review]

**Review by ranisalt — APPROVED** [review]

**Comment by conde2** [comment]
> > what is the context for this pull request? performance gain? resolving potential crashes? code cleanup?
> 
> It saves one virtual function call - which cannot be inlined because the function address cannot be determined in compile time, while `static_cast` is free - when we are sure that the conversion is valid.

@ranisalt 
But don't we need to call the dynamic_cast now? So instead of calling the virtual getPlayer we are calling the dynamic cast. Is this tradeoff worthy?

**Pull request merged**